Game/Resources/Shaders/billboard.fs

```diff
@@ -9,11 +13,44 @@ struct Billboard {
   float width;
   float height;
   sampler2D texture;
+  bool isSpritesheet;
+
+  vec4 color;
+
+  int XTiles;
+  int YTiles;
+  float speed;
+  int alignment_type;
+
 };
+
+
+
+
 uniform Billboard billboard;
 
+
 void main()
 {
-	vec4 texture_color = texture(billboard.texture, texCoord);
-	FragColor =  texture_color;
+	// Change to preprocessor directives
+	if(!billboard.isSpritesheet)
+	{
+		vec4 texture_color = texture(billboard.texture, texCoord)*billboard.color;
+		if(texture_color.a <0.1)
+		{
+			discard;
+		}
+		FragColor =  texture_color;
+	}
+
+	else
+	{
+		vec4 texture_color = texture(billboard.texture, frame)*billboard.color;
+		if(texture_color.a <0.1)
+		{
+			discard;
+		}
+		FragColor =  texture_color;
+	}
+
```

Comment:
Instead of duplicating the code block you can save the coordinates in a variable, something like this:

```
vec3 coordinates;
if(!billboard.isSpritesheet)
{
   coordinates = texCoord;
}else{
     coordinates = frame;
}
vec4 texture_color = texture(billboard.texture, coordinates)*billboard.color;
if(texture_color.a <0.1)
{
	discard;
}
     FragColor =  texture_color;

```

You can also use a tenary operator:

``` vec3 coordinates = !billboard.isSpritesheet?  texCoord : frame;```

Reply:
remember that in fragment shaders, the conditional "if" is very expensive, so sometimes is better to duplicate the code or to do a subrutine.

Reply:
yes, you are totally right